geps/gep-1713/index.md

```diff
@@ -0,0 +1,408 @@
+	Namespace *Namespace `json:"namespace,omitempty"`
+}
+```
+
+#### YAML
+
+The following example shows a `Gateway` with an HTTP listener and two child HTTPS `ListenerSets` with unique hostnames and certificates. 
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: foo
+    hostname: foo.com
+    protocol: HTTP
+    port: 80
+---
+apiVersion: gateway.networking.k8s.io/v1alpha1
+kind: ListenerSet
+metadata:
+  name: first-workload-listeners
+  labels:
+    gateway.networking.k8s.io/gateway.name: parent-gateway
+spec:
+  parentRef:
+    name: parent-gateway
+    kind: Gateway
+    group: gateway.networking.k8s.io
+  listeners:
+  - name: first
+    hostname: first.foo.com
+    protocol: HTTPS
+    port: 443
+    tls:
+      mode: Terminate
+      certificateRefs:
+      - kind: Secret
+        group: ""
+        name: first-workload-cert # Provisioned via HTTP01 challenge
+---
+apiVersion: gateway.networking.k8s.io/v1alpha1
+kind: ListenerSet
+metadata:
+  name: second-workload-listeners
+spec:
+  parentRef:
+    name: parent-gateway
+    kind: Gateway
+    group: gateway.networking.k8s.io
+  listeners:
+  - name: second
+    hostname: second.foo.com
+    protocol: HTTPS
+    port: 443
+    tls:
+      mode: Terminate
+      certificateRefs:
+      - kind: Secret
+        group: ""
+        name: second-workload-cert # Provisioned via HTTP01 challenge
+```
+### Semantics
+
+#### Gateway <> ListenerSet Handshake
+
+By default a `Gateway` will allow `ListenerSets` in the same namespace to be attached. Users can prevent this behaviour by configuring their `Gateway` to disallow any listener attachment:
+
+```
+apiVersion: gateway.networking.k8s.io/v1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  allowedListeners:
+  - from: None
+```
+
+#### Listeners
+
+####  `ListenerSet.spec.parentRef` is immutable
+
+A `ListenerSet` `MUST` be created with a `parentRef` Updating the `parentRef` of a ListenerSet `MUST` be rejected by the API.
+
+#### `ListenerSet` Default Labels
+
+In order for clients of the API to easily discover `Gateways` and their `ListenerSets` we propose a set of labels to be defaulted when creating `ListenerSets`. This only applies if the `parentRef.kind` is a `Gateway`
+
+- `gateway.networking.k8s.io/gateway.name` is equal to `spec.parentRef.name`
+- `gateway.networking.k8s.io/gateway.namespace` is equal to `spec.parentRef.namespace`
+
+This allows clients to perform API [lookups](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#list-and-watch-filtering) using label selectors.
+
+TBD: Can conditional defaulting be done with CEL?
+
+#### GatewaySpec.Listeners MinItems
+
+TBD: Do we want to make this change?
+
+Currently when creating a Gateway you must specify at least a single listener. With `ListenerSets` this opens the possibility of wanting to create a Gateway with no listeners.
```

Comment:
A potential use case for allowing no listeners is to apply different policies on different listeners in the same Gateway.
see: https://github.com/kubernetes-sigs/gateway-api/pull/3213#discussion_r1697308761

Reply:
> One advantage of making Listeners optional is that it would allow us to say "if you set AllowedListeners, you can't have any local Listeners and vice versa", which I think would make the merging decisions easier.

I'm a bit indifferent on this because it means existing Gateways with listeners can't opt into this feature

Reply:
If you have disjoint ownership of listeners and gateways, requiring one listener on the gateway is not really usable.

Imagine I have "Gateway Admin" and 2 devs creating ListenerSet A and B.

If we cannot have empty listeners on Gateway, the devs need to understand one was the "first" and modify the Gateway instead (which they don't have access to). If they later decide to remove A, they need to go find a new ~victim~ user to move their ListenerSet to the Gateway?

I think without this we will end up with people trying to make "dummy" listeners which is not ergonomic, and depending on the implementation may actually be a problem (i.e. I don't want to open up for any dummy listener for security/cost/etc purposes)

Reply:
> - What action do implementations take when they see a Gateway with no Listeners? Presumably the answer is "update the status to say the Gateway is okay as-is, even when it generates no configuration in any dataplane". Practically, that probably means that we'd need Accepted to be true, and Programmed to be false (since there's nothing to program)
...
> - Is the creation of a Gateway with no Listeners sufficient to trigger creation of dependent resources? I seem to recall people asking for Listener to be optional specifically to be able to start the creation of a Loadbalancer (which can take a while), before they know what ports the Gateway will need to listen on. I guess that whether this can succeed may depend on the loadbalancer in question - if it only requires allocation of an address for creation, rather than an address and ports, then maybe it could succeed. But regardless, we'll need to define what should happen in this case.

The Gateway has `spec.addresses`  - so I think `Accepted` and `Programmed` are still relevant in regarding that part of the spec.

> We would probably need to add a NoConfiguredListeners Reason to one of the Gateway-level Conditions (Programmed maybe?) or even an extra Condition.

Since I envision `Programmed` could be used to reflect `spec.addresses` status I don't think adding the `NoConfiguredListeners` reason should be done. 

I do think we can just leave an empty listener list in `status` - it's already `optional` in the spec. I think the semantics is clear what it means when `spec.listeners` is empty and so is `status.listeners`

> What do we do with the status.listeners` stanza? Presumably it stays empty, and we define an empty list there as meaning "no configured listeners" (since configured local listeners must be added regardless of their validity).

That sounds good

Reply:
I'll add a section about changes to Gateway mentioning this discussion

Reply:
section added - please revisit